src/MChestOnMap.java

```diff
@@ -0,0 +1,194 @@
+     */
+
+    /**
+     * Represents the chest's contents as a map of item names to quantities.
+     */
+    private Map<String, Integer> items;
+
+    /*
+     * Constructors -----------------------------------------------------------
+     */
+
+    /**
+     * No-argument constructor that initializes an empty chest.
+     *
+     * @ensures this = {}
+     */
+    public MChestOnMap() {
+        this.items = new Map2<>();
+    }
+
+    /*
+     * Kernel methods ---------------------------------------------------------
+     */
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void addItem(String item, int quantity) {
+        assert item != null : "Violation of: item is not null";
+        assert quantity > 0 : "Violation of: quantity > 0";
+
+        if (this.containsItem(item)) {
+            // If item exists, update the quantity
+            this.items.add(item, this.itemQuantity(item) + quantity);
+        } else {
+            // If item doesn't exist, add it to the chest
+            this.items.add(item, quantity);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Map.Pair<String, Integer> removeItem(String item, int quantity) {
+        assert item != null : "Violation of: item is not null";
+        assert quantity > 0 : "Violation of: quantity > 0";
+        assert this.containsItem(item) : "Violation of: item is in the chest";
+
+        int currentQuantity = this.itemQuantity(item);
+
+        if (currentQuantity <= quantity) {
+            Map.Pair<String, Integer> removedPair = this.items.remove(item);
+            return removedPair;
+        } else {
+            // Not gonna lie I didn't know how to bypass not making an instance
+            // of Map.Pair so that's why it's scuffed
+            this.items.remove(item);
+            this.items.add(item, currentQuantity - quantity);
+            Map.Pair<String, Integer> removedPair = this.items.remove(item);
+            return removedPair;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean containsItem(String item) {
+        assert item != null : "Violation of: item is not null";
+        return this.items.hasKey(item);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int itemQuantity(String item) {
+        assert item != null : "Violation of: item is not null";
+        if (this.items.hasKey(item)) {
+            return this.items.value(item);
+        } else {
+            return 0;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isFull() {
+        final int thousand = 1000;
+        return this.totalItems() >= thousand;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void clear() {
+        this.items = new Map2<>(); // Reset the chest to an empty state
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
```

Comment:
The common methods should be in the abstract class.

Reply:
Unless they already are, and you're overriding them here for performance reasons.